From b86cb45fde94a01aed192dc3eb0be33a276d95be Mon Sep 17 00:00:00 2001
From: martinbaker <ax87438@martinb.com>
Date: Sun, 4 Dec 2016 15:35:20 +0000
Subject: [PATCH 10/10] group conversions patch

---
 src/input/gpresent.input | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

diff --git a/src/input/gpresent.input b/src/input/gpresent.input
index 2c41612..5dc1d88 100644
--- a/src/input/gpresent.input
+++ b/src/input/gpresent.input
@@ -94,5 +94,32 @@ testEquals("expected2", "t9")
 t10 := quotient(t8,[[1]])
 testEquals("expected2", "t10")
 
+-- conversions to and from PermutationGroup
+
+testcase "PermutationGroup conversions cyclic group"
+
+c5a := cyclicGroup(5)$PermutationGroupExamples
+c5b := c5a::GroupPresentation
+c5c := toPermutationIfCan(c5b) :: PermutationGroup(Integer)
+
+testEquals("c5a", "c5c")
+
+testcase "PermutationGroup conversions dihedral group"
+
+d3a := dihedralGroup(3)$PermutationGroupExamples
+d3b := d3a::GroupPresentation
+d3c := toPermutationIfCan(d3b) :: PermutationGroup(Integer)
+
+testEquals("order(d3a)", "order(d3c)")
+
+testcase "PermutationGroup conversions containing inverse elements"
+
+d3a := dihedralGroup(3)$GroupPresentation
+d3b := toPermutationIfCan(d3a) :: PermutationGroup(Integer)
+d3c := groupPresentation([1,2],[[1,1,1],[2,2],[-2,-1,-2,-1]])
+d3d := toPermutationIfCan(d3c) :: PermutationGroup(Integer)
+
+testEquals("d3b", "d3d")
+
 statistics()
 
-- 
2.1.4

