From 53e5fe25f7f13b48caa4a57b1592842d71c04398 Mon Sep 17 00:00:00 2001
From: martinbaker <ax87438@martinb.com>
Date: Wed, 9 Nov 2016 09:49:10 +0000
Subject: [PATCH 7/7] regression tests for simplify improvements

---
 src/input/gpresent.input | 31 ++++++++++++++++++++++++++++---
 1 file changed, 28 insertions(+), 3 deletions(-)

diff --git a/src/input/gpresent.input b/src/input/gpresent.input
index e03cc74..2c41612 100644
--- a/src/input/gpresent.input
+++ b/src/input/gpresent.input
@@ -46,6 +46,31 @@ t6 := simplify(square)
 expected := groupPresentation([1,2,4,5],[[2,-1,-5,4]])
 testEquals("expected", "t6")
 
+testcase "simplification - remove duplicate rule"
+x := groupPresentation([1,2,3],[[1,2,3],[1,2,3]])
+t6 := simplify(x)
+expected := groupPresentation([1,2,3],[[1,2,3]])
+testEquals("expected", "t6")
+y := groupPresentation([1,2,3],[[1,2,3],[-3,-2,-1]])
+t7 := simplify(y)
+testEquals("expected", "t7")
+z := groupPresentation([1,2,3],[[1,2,3],[2,3,1]])
+t8 := simplify(z)
+testEquals("expected", "t8")
+a := groupPresentation([1,2,3],[[1,2,3],[2,3,1],[-3,-2,-1]])
+t9 := simplify(a)
+testEquals("expected", "t9")
+b := groupPresentation([1,2,3,4,5,6],[[1,2,3],[4,5,6],[4,5,6]])
+t10 := simplify(b)
+expected2 := groupPresentation([1,2,3,4,5,6],[[1,2,3],[4,5,6]])
+testEquals("expected2", "t10")
+c := groupPresentation([1,2,3,4,5,6],[[4,5,6],[1,2,3],[4,5,6]])
+t11 := simplify(c)
+testEquals("expected2", "t11")
+d := groupPresentation([1,2,3,4,5,6],[[4,5,6],[4,5,6],[1,2,3]])
+t12 := simplify(d)
+testEquals("expected2", "t12")
+
 -- more complicated simplifications
 
 testcase "complicated simplification"
@@ -56,7 +81,7 @@ t7 := simplify(c1)
 expected := groupPresentation([],[])
 testEquals("expected", "t7")
 
--- direct product and subgroup
+-- direct product and quotient
 
 testcase "direct product"
 a := groupPresentation([1],[])
@@ -64,9 +89,9 @@ t8 := directProduct(a,a)
 expected := groupPresentation([1,2],[[1,2,1,2]])
 testEquals("expected", "t8")
 expected2 := groupPresentation([2],[[2,2]])
-t9 := subgroup(t8,[1])
+t9 := quotient(t8,[1])
 testEquals("expected2", "t9")
-t10 := subgroup(t8,[[1]])
+t10 := quotient(t8,[[1]])
 testEquals("expected2", "t10")
 
 statistics()
-- 
2.1.4

