From d37fa3866267f945fca851d65c226edd6ae6b1ad Mon Sep 17 00:00:00 2001
From: martinbaker <ax87438@martinb.com>
Date: Sun, 25 Sep 2016 16:16:50 +0100
Subject: [PATCH 7/7] fix regression bug gpresent to OutputForm

---
 src/algebra/gpresent.spad | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/src/algebra/gpresent.spad b/src/algebra/gpresent.spad
index 962b9ee..3ced5a0 100644
--- a/src/algebra/gpresent.spad
+++ b/src/algebra/gpresent.spad
@@ -546,19 +546,19 @@ GroupPresentation() : Exports == Impl where
        rs : List(List(Integer)) := s.rels
        rels1 : List(OutputForm) := empty()$List(OutputForm)
        for r in rs repeat
-           eleout : OutputForm := outputForm(" ")
-           seperator : OutputForm := outputForm(" ")
+           eleout : OutputForm := message(" ")
+           seperator : OutputForm := message(" ")
            for ele in r repeat
                newterm : OutputForm := genName(ele)
                eleout := hconcat([eleout, seperator, newterm])$OutputForm
-               seperator := outputForm("*")
+               seperator := message("*")
            rels1 := concat(rels1, eleout)
        --print("output rules : " << rels1)
-       g : OutputForm := outputForm(" ")
+       g : OutputForm := message(" ")
        if #gens1 > 0 then g := blankSeparate(gens1)
-       r : OutputForm := outputForm(" ")
+       r : OutputForm := message(" ")
        if #rels1 > 0 then r := commaSeparate(rels1)
-       hconcat([outputForm("<"), g, outputForm(" | "), r, outputForm(">")])
+       hconcat([message("<"), g, message(" | "), r, message(">")])
 
 )if false
 \section{Homology}
-- 
2.1.4

